libs/s25main/mapGenerator/HeadQuarters.h

```diff
@@ -38,9 +38,14 @@ namespace rttr { namespace mapGenerator {
     std::vector<MapPoint> FindLargestConnectedArea(const Map& map);
 
     /**
-     * Finds the most suitable position for a HQ in the specified area of the map. To find the most suitable position
-     * for the entire map just leave the area empty. The resulting HQ positions are sorted by quality (highest quality
-     * first). Good HQ positions are positions which are far away from other HQs and in a widely buildable area.
+     * Finds suitable positions for a HQ in the specified area of the map. The resulting HQ positions are sorted by quality.
+     * To find suitable positions, this function does:
+     * 1. look for points within a buildable area of radius 2 (min. req. of HQ)
+     * 2. if no such point exists return empty
+     * 3. sort all those points by their distance to existing HQs
+     * 4. filter out points within minimum distance to existing HQs
+     * 5. if no remaining points just return result of 3.
+     * 6. sort remaining points by how close they're to desired mountain distance
```

Comment:
Ok, prior this (point 6) was (good case only): 

- Choose points which closest match desired mountain distance
- Sort by distance to other HQs (descending)

Correct? Asking for comparison and also as the `minObstacleDistance` is gone which (I think) usually placed the HQs in radius 4 (not 2), but not sure if that mattered in practice.
Anyway the new algorithm is good too, just 2 issues:

1. The sorting in point 3 (distance to other HQs) now only matters for the fallback point 5 (min distance to other HQs not possible), so it should be moved into that to avoid the work
1. I expect that there are many points which are "same close" to desired mountain distance, correct?
Would it make sense to sort those by HQ distance? Basically similar to what it was before, where quality was always related to HQ distance. For this you can either include the HQ distance as a tie breaker in the sort functor or do the sorting in point 3 as you currently do and use `std::stable_sort` (which allocates memory)

Minor notes for the future:
- I would not _require_ to repeat the points in the code, but it's nice to have.
- Point 2 could be removed from the description without changing the others, I'd treat that as an optimization because the sorting and filtering also works for "no points". Makes the description shorter (only reason I mention that)

Reply:
> Ok, prior this (point 6) was (good case only):
> Choose points which closest match desired mountain distance
> Sort by distance to other HQs (descending)
> Correct? 

Yes, that's right. I removed the whole thing around `minObstacleDistance` which happened to be a left-over before this logic changed. 2 is the clamped minimum anyway, so that wouldn't change much. 

> The sorting in point 3 (distance to other HQs) now only matters for the fallback point 5 (min distance to other HQs not possible), so it should be moved into that to avoid the work

I assumed that would keep the order in case objects have the same value when sorting for mountain distance afterwards. Obviously, I've been wrong, thanks for mentioning `std::stable_sort`, I think that's the way to go in this case. 

> I would not require to repeat the points in the code, but it's nice to have.

I felt like those points would help a dev looking through the code. Probably better if the code reflects those points directly, rather than repeating the comments. 

> Point 2 could be removed from the description without changing the others, I'd treat that as an optimization because the sorting and filtering also works for "no points". Makes the description shorter (only reason I mention that)

That was intended as an optimization to avoid calculating `distanceToOtherHqs` if that makes sense. Happy to remove it from the description anyway as it's kind of obvious.